components/director/internal/tenantfetchersvc/subscriber.go

```diff
@@ -0,0 +1,124 @@
+}
+
+type subscriptionFunc func(ctx context.Context, tenantSubscriptionRequest *TenantSubscriptionRequest, region string) error
+
+type sliceMutationFunc func([]string, string) []string
+
+type subscriber struct {
+	provisioner TenantProvisioner
+	RuntimeService
+	RegionLabelKey                string
+	SubscriptionConsumerLabelKey  string
+	ConsumerSubaccountIDsLabelKey string
+}
+
+// NewSubscriber creates new subscriber
+func NewSubscriber(provisioner TenantProvisioner, service RuntimeService, regionLabelKey, subscriptionConsumerLabelKey, consumerSubaccountIDsLabelKey string) *subscriber {
+	return &subscriber{
+		provisioner:                   provisioner,
+		RuntimeService:                service,
+		RegionLabelKey:                regionLabelKey,
+		SubscriptionConsumerLabelKey:  subscriptionConsumerLabelKey,
+		ConsumerSubaccountIDsLabelKey: consumerSubaccountIDsLabelKey,
+	}
+}
+
+// Subscribe subscribes tenant to runtime. If the tenant does not exist it will be created
+func (s *subscriber) Subscribe(ctx context.Context, tenantSubscriptionRequest *TenantSubscriptionRequest, region string) error {
+	if err := s.provisioner.ProvisionTenants(ctx, tenantSubscriptionRequest, region); err != nil {
+		return err
+	}
+
+	return s.applyRuntimesSubscriptionChange(ctx, tenantSubscriptionRequest.SubscriptionConsumerID, tenantSubscriptionRequest.SubaccountTenantID, region, addElement)
+}
+
+// Unsubscribe unsubscribes tenant from runtime.
+func (s *subscriber) Unsubscribe(ctx context.Context, tenantSubscriptionRequest *TenantSubscriptionRequest, region string) error {
+	return s.applyRuntimesSubscriptionChange(ctx, tenantSubscriptionRequest.SubscriptionConsumerID, tenantSubscriptionRequest.SubaccountTenantID, region, removeElement)
+}
+
+func (s *subscriber) applyRuntimesSubscriptionChange(ctx context.Context, subscriptionConsumerID, subaccountTenantID, region string, mutateSliceFunc sliceMutationFunc) error {
+	filters := []*labelfilter.LabelFilter{
+		labelfilter.NewForKeyWithQuery(s.SubscriptionConsumerLabelKey, fmt.Sprintf("\"%s\"", subscriptionConsumerID)),
+		labelfilter.NewForKeyWithQuery(s.RegionLabelKey, fmt.Sprintf("\"%s\"", region)),
+	}
+
+	runtimes, err := s.ListByFiltersGlobal(ctx, filters)
+	if err != nil {
+		if apperrors.IsNotFoundError(err) {
+			return nil
+		}
+
+		return errors.Wrap(err, fmt.Sprintf("Failed to get runtimes for labels %s: %s and %s: %s", s.RegionLabelKey, region, s.SubscriptionConsumerLabelKey, subscriptionConsumerID))
+	}
+
+	for _, runtime := range runtimes {
+		ctx = tenant.SaveToContext(ctx, runtime.Tenant, "")
+
+		labelOldValue := make([]string, 0)
+		label, err := s.GetLabel(ctx, runtime.ID, s.ConsumerSubaccountIDsLabelKey)
+		if err != nil {
+			if !apperrors.IsNotFoundError(err) {
+				return errors.Wrap(err, fmt.Sprintf("Failed to get label for runtime with id: %s and key: %s", runtime.ID, s.ConsumerSubaccountIDsLabelKey))
+			}
+			// if the error is not found, do nothing and continue
```

Comment:
If you come in this comment you will anyway modify the labels of the runtime to empty array because the code continues to the `SetLabel` invocation. Perhaps `continue` should be added.